src/index.c

```diff
@@ -1372,14 +1391,72 @@ static int OI_SkipTo(void *ctx, t_docId docId, RSIndexResult **hit) {
   return INDEXREAD_OK;
 }
 
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+static int OI_SkipTo_O(void *ctx, t_docId docId, RSIndexResult **hit) {
+  OptionalMatchContext *nc = ctx;
+
+  bool found = false;
+
+  if (nc->lastDocId > nc->maxDocId) {
+    IITER_SET_EOF(nc->wcii);
+    IITER_SET_EOF(&nc->base);
+    return INDEXREAD_EOF;
+  }
+
+  if (docId == 0) {
+    // No doc was read yet - read the first doc
+    return nc->base.Read(ctx, hit);
+  }
+
+  int rc;
+
+  if (docId == nc->nextRealId) {
+    found = true;
+    nc->base.current = nc->child->current;
+  } else if (docId > nc->nextRealId) {
+    rc = nc->child->SkipTo(nc->child->ctx, docId, &nc->base.current);
+    if (rc == INDEXREAD_OK) {
+      found = true;
+    }
+    if (nc->base.current) {
+      nc->nextRealId = nc->base.current->docId;
+    }
+  }
+
+  // Promote the wildcard iterator to the requested docId if the docId
+  RSIndexResult *wcii_res = NULL;
+  if (docId > nc->wcii->LastDocId(nc->wcii->ctx)) {
+    rc = nc->wcii->SkipTo(nc->wcii->ctx, docId, &wcii_res);
+    if (rc == INDEXREAD_EOF) {
+      IITER_SET_EOF(&nc->base);
+      return INDEXREAD_EOF;
+    } else if (rc == INDEXREAD_NOTFOUND) {
+      // This doc-id was deleted
+      return INDEXREAD_NOTFOUND;
+    }
```

Comment:
Can any of these cases happen when we previously set `found = true`?

Reply:
Yes, if the found doc was deleted, and the GC did not update the inverted index yet

Reply:
I'll add a test with deleted docs (wanted to do it anyway..)